非同期処理(JavaScript)を実装しました

yumiya-g opened the pull request.

yumiya-g committed:
Add: `node-sqlite3`をインストールしました

yumiya-g committed:
Add: コールバック関数で、エラーあり/なしのプログラムを実装しました

yumiya-g committed:
Fix: コールバック関数でエラーなしの処理で、自動採番されたIDを表示しました

yumiya-g committed:
Add: Promiseで、エラーなしのプログラムを実装しました

yumiya-g committed:
Fix: Promiseでエラーなしのプログラムを実装しました

- エラーあり/なしでファイルを分割して共通化する処理に変更

yumiya-g committed:
Add: Asyncでエラーなしのプログラムを実装しました

yumiya-g committed:
Rename: Async/awaitでも使用するため、分割したファイル名を変更しました

yumiya-g committed:
Add: Asyncでエラーあり/なしのプログラムを実装しました

- データベースからの出力は`finally`メソッド内で実行

cafedomancer commented on 03.asynchronous/callback-has-errors.js:
no-errors と has-errors は品詞的に対称的じゃなさそうです。また、ファイル名は名詞形になるようにしてください。現状のファイル名だと文の形になっているように見えます。

yumiya-g replied:
エラー有り無しでファイルを分割していたのですが、修正の際に1ファイルにまとめてファイル名を修正しました🙏

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+
+const db = new sqlite3.Database(":memory:");
+
+const createTableQuery = `CREATE TABLE IF NOT EXISTS books (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    title TEXT NOT NULL UNIQUE
+  )`;
```
テンプレート文字列で改行し、かつインデントすると以下のように余分な空白が含まれてしまいます。

```
CREATE TABLE IF NOT EXISTS books (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    title TEXT NOT NULL UNIQUE
  )
```

それと、今回のプログラムではインメモリデータベースを使っているので、プログラムの実行時点でテーブルは存在する可能性はなく、`IF NOT EIXSTS` をつけても特に効果がないです。

また、質問なのですが、この SQL は変数に格納されていて、他の SQL は変数に格納されていないのは理由があるんでしょうか？ 一貫性のない点が気になっています。

yumiya-g replied:
> テンプレート文字列で改行し、かつインデントすると以下のように余分な空白が含まれてしまいます。

SQL文は改行をせず1行に修正しました💡

> IF NOT EIXSTS をつけても特に効果がないです。

こちらは削除をしました。

>  一貫性のない点が気になっています。

単純にチェック漏れのため、変数にSQL文を格納していませんでした😓

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+    title TEXT NOT NULL UNIQUE
+  )`;
+
+db.run(createTableQuery, function (err) {
```
function 式とアロー関数の使い分けを行ってください。その関数の `this` が関数内部で必要ない場合はアロー関数を使ってください。

yumiya-g replied:
標準出力をする箇所でfunction式を使い、その他はアロー関数に変更をしました。

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+  )`;
+
+db.run(createTableQuery, function (err) {
+  if (err) {
+    console.log(err.message);
+  } else {
```
今回の課題では、エラーを起こさない想定の部分のエラーハンドリングは不要です。

yumiya-g replied:
エラー無しの処理では、エラーハンドリングを除外しました💡

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+    console.log(err.message);
+  } else {
+    db.run(
+      "insert into books(title) values(?)",
```
SQL のキーワード部分は大文字で記述してください。また、CREATE TABLE とスタイルを合わせるのであれば `(` の前にスペースを入れてください。

yumiya-g replied:
SQL文のキーワードを大文字に変更し、変数に格納して使い回せるように修正しました🙏

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+                console.log(`ID: ${this.lastID}`);
+                db.each(
+                  "select * from books",
+                  (err, row) => {
```
使われていない変数 (`err`) はそれが分かる命名規則を適用してください。

yumiya-g replied:
使用されていない変数を、`_`（アンダースコアのみ）か、`_err`にするかで検討しましたが、後者の`_err`を適応しました。

cafedomancer commented on 03.asynchronous/callback-no-errors.js:
```diff
@@ -0,0 +1,50 @@
+                  function (err) {
+                    if (err) {
+                      console.log(err.message);
+                    } else {
+                      db.close();
+                    }
```
エラーがあってもなくても close の処理を実行する必要があると思います。

yumiya-g replied:
`each`メソッドの完了コールバック関数にてclose を実行する処理に修正しました。

https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

cafedomancer commented on 03.asynchronous/callback-has-errors.js:
```diff
@@ -0,0 +1,71 @@
+import sqlite3 from "sqlite3";
+import timers from "timers/promises";
```
Node.js のライブラリ、NPM パッケージ、自分の定義したものの順に import を記述してください。

yumiya-g replied:
> Node.js のライブラリ、NPM パッケージ、自分の定義したものの順に import を記述してください。

ご指摘いただいた順に記述をしなおしました。
ちなみにこの順番で記述するのは、どのような理由があるのか気になりましたので、ご教示いただけますと幸いです🙏

cafedomancer commented on 03.asynchronous/callback-has-errors.js:
```diff
@@ -0,0 +1,71 @@
+                  if (err) {
+                    console.log(err.message);
+                  } else {
+                    db.close();
+                  }
+                },
+              );
+            }
+          });
+        }
+      },
+    );
+  }
+});
+
+await timers.setTimeout(300);
+
+db.run(
+  "INSERT INTO books(title) VALUES(?)",
+  "初めてのJavaScript",
+  function (err) {
+    if (err) {
+      console.log(err.message);
+    } else {
+      db.each(
+        "SELECT * FROM booksss ",
+        (err, row) => {
+          console.log(`${row.id}, title:${row.title}`);
+        },
+        function (err) {
+          if (err) {
+            console.log(err.message);
+          } else {
+            db.close();
+          }
+        },
+      );
+    }
+  },
+);
```
実行結果は要件を満たしているのですが、プログラムの流れが要件と対応していないです。コールバックのプログラムと同じように、Docs の 1 から 4 の手順を表現するようなプログラムにしてください。

yumiya-g replied:
エラーありの処理を途中で終了させず、多段のネストになるよう実装しなおしました🙏

cafedomancer commented on 03.asynchronous/promise-no-errors.js:
```diff
@@ -0,0 +1,40 @@
+    console.log(error);
+  })
+  .finally(() => {
+    noError.db.each(
+      "select * from books",
+      (err, row) => {
+        console.log(`ID: ${row.id}, タイトル: ${row.title}`);
+      },
+      function (err) {
+        if (err) {
+          console.log(err.message);
+        } else {
+          noError.db.close();
+        }
+      },
+    );
```
each も非同期処理を行う node-sqlite3 の関数なのに Promise として扱えるようになっていません。close についても同様です。

yumiya-g replied:
node-sqlite3 の関数 を`new Promise`でラップして扱うように修正しました🙏

cafedomancer commented on 03.asynchronous/registerBooks.js:
```diff
@@ -0,0 +1,21 @@
+    title TEXT NOT NULL UNIQUE
+  )`;
+
+export function registerBooks(query, params = []) {
+  return new Promise((resolve, reject) => {
+    db.run(query, params, function (err) {
+      if (err) {
+        reject(err);
+      } else {
+        console.log(`ID: ${this.lastID}`);
+        resolve(this.lastID);
+      }
+    });
+  });
+}
```
独自の抽象度の関数は定義しないでください。コールバックのプログラムで使用した node-sqlite3 の関数に対応する関数を同じ抽象度で定義してください。

yumiya-g replied:
こちらの`registerBooks`関数は削除して、node-sqlite3 の関数を`Promise`でラップするだけの関数に変更しました。

cafedomancer commented on 03.asynchronous/registerBooks.js:
```diff
@@ -0,0 +1,21 @@
+
+export function registerBooks(query, params = []) {
+  return new Promise((resolve, reject) => {
+    db.run(query, params, function (err) {
```
関数がその外側のスコープの変数 (`db`) に依存していると関数の再利用性が下ります。例えば、ここで定義している関数を別の `sqlite3.Database` オブジェクトと組み合わせて使おうと思ってもそうできません。

yumiya-g replied:
> 関数がその外側のスコープの変数 (db) に依存していると関数の再利用性が下ります。

`Promise`でラップした関数の中から、スコープ外の変数を直接呼び出すのではなく、関数に引数として変数（db）を渡す処理に修正しました。

cafedomancer commented on 03.asynchronous/promise-no-errors.js:
```diff
@@ -0,0 +1,40 @@
+
+noError
+  .registerBooks(noError.createTableQuery)
+  .then(() => {
+    return noError.registerBooks(
+      "insert into books(title) values(?)",
+      "スラスラ読める JavaScriptふりがなプログラミング",
+    );
+  })
```
アロー関数の中身が return しかないときは簡潔文体を使ってください。

yumiya-g replied:
アロー関数の戻り値が`return`の時に、簡潔文体になるよう調整をしました💡

cafedomancer commented on 03.asynchronous/promise-no-errors.js:
```diff
@@ -0,0 +1,40 @@
+  .catch((error) => {
+    console.log(error);
+  })
+  .finally(() => {
+    noError.db.each(
+      "select * from books",
+      (err, row) => {
+        console.log(`ID: ${row.id}, タイトル: ${row.title}`);
+      },
+      function (err) {
+        if (err) {
+          console.log(err.message);
+        } else {
+          noError.db.close();
+        }
+      },
+    );
+  });
```
each は finally で処理する必要のある処理内容 (エラーなどが起きても必ず実行する必要のある処理) ではないと思います。

yumiya-g replied:
`each`を`finally`メソッドから除外しました🙏

yumiya-g committed:
Fix: SQL文を修正して、`INSERT`文と`SELECT`文を変数に代入しました

- `IF NOT EXISTS`を削除
- SQL文を大文字に修正

yumiya-g committed:
Fix: エラーなしコールバック関数のエラーハンドリングを調整しました

- エラーハンドリング処理を削除
- データベースの出力をアロー関数に変更
- 使用していない変数`err`を`_err`に変更
- データベースのクローズ処理を条件分岐せず実行するよう修正

yumiya-g committed:
Fix: コールバック関数を修正しました

- `import`順を見直し
- エラーありの処理を1ファイルにまとめる
  - `timers.setTimeout()`で待ち時間を追加
  - エラーなし処理用のデータベースを作成
  - SQL文を変数に代入

yumiya-g committed:
Rename: コールバック関数用のファイル名を変更しました

- エラー処理用のファイルは削除

yumiya-g committed:
Fix: Promiseエラーなしの修正を行いました

- resisterBooks関数を外す
- SQL文を修正して変数に代入
- function式とアロー関数の使い分け
- `finally`ではなく`then`でデータベースを閉じる

yumiya-g committed:
Add: Promiseでエラーありの修正を行いました

- `timers/promises`APIを追加
- エラーを発生させるSQL文を変数に代入
- エラーなしのプログラム内のアロー関数を調整
- データベース接続を閉じる処理を`finally`メソッドに修正

yumiya-g committed:
Fix: コールバック関数のデータベースの接続処理を修正しました

yumiya-g committed:
Fix: Promiseを使った処理を修正しました

- データベースへ接続する変数を`db`に統一
- 各メソッドに引数`db`を渡して、スコープ外の変数を使用しない
- データベースのクローズ処理を`finally`メソッドで実行する

yumiya-g committed:
Renama: Promise用のファイル名を変更しました

- Promiseのエラー専用ファイルは削除

yumiya-g committed:
Fix: Promiseの`then`メソッドに渡すオブジェクトのキー名を修正しました

yumiya-g committed:
Fix: Async/awaitエラーなしの修正を行いました

- resisterBooks関数を使用しない
- データベース作成とデータ追加、データベースを閉じる関数を作成

yumiya-g committed:
Fix: Async/awaitエラーなしの調整を行いました

- `asyncNoError`関数の引数に変数`db`を追加
- `each`メソッドの完了コールバック関数に`resolve`を登録して状態を変更
- `close`メソッドの完了コールバック関数に`resolve`を登録して処理を終了

yumiya-g committed:
Add: Async/awaitでエラーありの修正を行いました

- エラーを発生させるSQL文を変数に代入
- `asyncError`関数の引数に変数dbを追加
- `try...catch`構文でエラー発生時のメッセージを出力

yumiya-g committed:
Rename: Async/await用のファイル名を変更しました

- 不要ファイルを削除
- エラーなしの関数`asyncNoError`内の引数名を修正

yumiya-g committed:
Fix: 重複していた`invalid_promise`関数を削除しました

yumiya-g committed:
Fix: コールバック関数の処理を修正しました

- `this`を使用しない関数をアロー関数に変更
- `return`しか返さないアロー関数を簡潔文体に変更
- データベースを閉じる処理を、`each`メソッドのcallback内で実行

yumiya-g committed:
Fix: Promiseのアロー関数を調整しました

- `this`を使用しない関数をアロー関数に変更
- `return`しか返さないアロー関数を簡潔文体に変更

yumiya-g committed:
Fix: Async/awaitのアロー関数を調整しました

- `return`しか返さないアロー関数を簡潔文体に変更

yumiya-g committed:
Fix: `node-sqlite3`の関数を`Promise`でラップする関数を作成しました

yumiya-g committed:
Fix: 共通する変数や関数を`utils.js`に分割しました

cafedomancer commented on 03.asynchronous/utils.js:
utils という名前では何が定義されているファイルなのか分からないです。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,56 @@
+import timers from "timers/promises";
+import sqlite3 from "sqlite3";
+import * as utils from "./utils.js";
```
質問ですが、これだけなぜ `as` を使って import しているんでしょうか？

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,56 @@
+
+let db = new sqlite3.Database(":memory:");
+
+db.run(utils.createTableQuery, () => {
+  db.run(
+    utils.insertTableQuery,
+    "スラスラ読める JavaScriptふりがなプログラミング",
+    function () {
+      console.log(`ID: ${this.lastID}`);
+      db.run(utils.insertTableQuery, "初めてのJavaScript", function () {
+        console.log(`ID: ${this.lastID}`);
+        db.each(
+          utils.selectTableQuery,
+          (_err, row) => console.log(`ID: ${row.id}, タイトル: ${row.title}`),
+          () => db.close(),
+        );
+      });
+    },
+  );
+});
```
要件にあるテーブルを削除する処理がないです。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,56 @@
+        console.log(`ID: ${this.lastID}`);
+        db.each(
+          utils.selectTableQuery,
+          (_err, row) => console.log(`ID: ${row.id}, タイトル: ${row.title}`),
```
アロー関数の戻り値がないときはブロック文体を使って書いてください。戻り値がないことを明確にするためです。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,56 @@
+    "スラスラ読める JavaScriptふりがなプログラミング",
+    (err) => {
+      if (err) {
+        console.log(err.message);
```
標準エラー出力にエラーメッセージを出力してください。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,56 @@
+    (err) => {
+      if (err) {
+        console.log(err.message);
+        db.run(utils.insertTableQuery, null, (err) => {
+          if (err) {
+            console.log(err.message);
+            db.each(utils.selectTableWrongQuery, (err, _row) => {
+              if (err) {
+                console.log(err.message);
+              } else {
+                console.log(_row);
+              }
+              () => db.close();
+            });
+          } else {
+            console.log("エラーなし");
+          }
+        });
```
後続の処理は前の処理のエラーの有無にかかわらず実行されるようにしてください。これだとエラーがあった場合にしか実行されないです。

cafedomancer commented on 03.asynchronous/utils.js:
```diff
@@ -0,0 +1,84 @@
+export const createTableQuery = `CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)`;
```
文字列中に変数を埋め込んでないのにテンプレート文字列が使われています。

cafedomancer commented on 03.asynchronous/utils.js:
```diff
@@ -0,0 +1,84 @@
+export const createTableQuery = `CREATE TABLE books (id INTEGER PRIMARY KEY AUTOINCREMENT, title TEXT NOT NULL UNIQUE)`;
+export const insertTableQuery = `INSERT INTO books (title) VALUES(?)`;
```
スタイルが整っていないです。

```suggestion
export const insertTableQuery = `INSERT INTO books (title) VALUES (?)`;
```

cafedomancer commented on 03.asynchronous/utils.js:
```diff
@@ -0,0 +1,84 @@
+export const insertFirstBookError = (db) =>
+  new Promise((_, reject) =>
+    db.run(
+      insertTableWrongQuery,
+      "スラスラ読める JavaScriptふりがなプログラミング",
+      function (err) {
+        if (err) {
+          console.log(err.message);
+          reject(db);
+        } else {
+          _(this);
+        }
+      },
+    ),
+  );
+
+export const insertSecondBookError = (db) =>
+  new Promise((_, reject) =>
+    db.run(insertTableQuery, null, function (err) {
+      if (err) {
+        console.log(err.message);
+        reject(db);
+      } else {
+        _(this);
+      }
+    }),
+  );
+
+export const displayBooksError = (db) =>
+  new Promise((_, reject) =>
+    db.each(selectTableWrongQuery, (err, _row) => {
+      if (err) {
+        console.log(err.message);
+        reject(db);
+      } else {
+        console.log(`ID: ${_row.id}, タイトル: ${_row.title}`);
+        _(this);
+      }
+    }),
+  );
```
「独自の抽象度の関数を定義するのではなく、node-sqlite3 の関数 (run, get, all など) を Promise ベースにするような関数を定義してください。」という注意点に引っかかっています。

yumiya-g committed:
Rename: `utils.js`を、変数用と関数用のファイルに分割しました

- queries.js:SQL文をダブルクォートで囲み、スタイルを調整

yumiya-g committed:
Fix: 分割したファイルからインポートする変数名を指定して、リスト化しました

yumiya-g committed:
Fix: 戻り値がないアロー関数をブロック文体に変更しました

yumiya-g committed:
Fix: エラーメッセージを標準エラー出力で表示するように修正しました

yumiya-g committed:
Fix: データベースに登録された内容を出力後、テーブルを削除して切断する処理を追加しました

- データベースを削除するSQL文を`queries.js`に追加

yumiya-g committed:
Fix: 前の処理でエラーの有無に関わらず実行されるよう修正しました

yumiya-g committed:
Change: SQL文を別ファイルに分割したので、importする処理を追加しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,75 @@
+      if (err) {
+        console.error(err.message);
+      } else {
+        console.log("エラーなし");
```
「エラーなし」と出力するのではなく、エラーなしのプログラムと同様に正常時の処理を書いたほうが良いと思います。

yumiya-g replied:
正常系のコードを追記しました🙏

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,75 @@
+      } else {
+        console.log("エラーなし");
+      }
+
```
エラーなしのプログラムには処理間に空行がないです。あってもなくてもよいですがスタイルを一貫させてください。

yumiya-g replied:
空行がないようにスタイルを調整しました。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,75 @@
+          console.log("エラーなし");
+        }
+
+        db.each(selectTableWrongQuery, (err, _row) => {
```
`_row` は使われている変数なのだから `_` を付けてはダメです。

yumiya-g replied:
こちらは`_`を消し忘れていました😭
また、今回の修正でエラーありのプログラム内に正常系の処理を追記して`_`を削除しました。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        db.each(selectTableWrongQuery, (err, _row) => {
+          if (err) {
+            console.error(err.message);
+          } else {
+            console.log(_row);
+          }
+          () => {
+            db.run(deleteTableQuery, () => {
+              db.close();
+            });
+          };
```
何かおかしいです。削除の処理は実行されていないですよ。

yumiya-g replied:
`each`メソッドのコールバック関数内ではなく、完了関数として削除を実行しているつもりでした。。
今回の修正では`all`メソッドに変更をして、コールバック関数内で削除を実行しています🙏

cafedomancer commented on 03.asynchronous/queries.js:
```diff
@@ -0,0 +1,16 @@
+const selectTableQuery = "SELECT * FROM books";
+const insertTableWrongQuery = "INSERT INTO bookss (title) VALUES (?)";
+const selectTableWrongQuery = "SELECT hoge FROM bbbooks";
+const deleteTableQuery = "DROP TABLE IF EXISTS books";
```
削除されないとおかしいのだから `IF EXISTS` を付けてはダメだと思います。

yumiya-g replied:
テーブルの有無に関わらず`books`を削除するため、`IF EXISTS`を削除しました。

cafedomancer commented on 03.asynchronous/queries.js:
```diff
@@ -0,0 +1,16 @@
+const selectTableWrongQuery = "SELECT hoge FROM bbbooks";
+const deleteTableQuery = "DROP TABLE IF EXISTS books";
+
+export {
+  createTableQuery,
+  insertTableQuery,
+  selectTableQuery,
+  insertTableWrongQuery,
+  selectTableWrongQuery,
+  deleteTableQuery,
+};
```
まとめて export するのでも良いですが、各宣言の前に export を付けるほうが export されていることが分かりやすいと思います。これだと最下部まで行かないと export されていることが分からないからです。

yumiya-g replied:
`export`する変数や関数の数が多くなるとわかりづらくなりますので、各変数にexport宣言をつけるように修正しました🙏

yumiya-g committed:
Fix: テーブルを削除するSQL文と、各変数の`export`宣言の方法を修正しました

yumiya-g committed:
Fix: エラー処理内の正常系処理の修正を行いました

- 正常系の処理でID出力を行うため、コールバック関数をアロー関数から`function`に変更
- 変数`_row`は使用されているので`row`に修正

yumiya-g committed:
Fix: エラー処理内のデータベース接続処理を修正しました

- データベースが空の場合を考慮して`each`メソッドから`all`メソッドに修正
- 合わせてエラーなしの出力処理も`all`メソッドに変更

yumiya-g committed:
Fix: コールバック関数を修正しました

- データの出力処理を`all`メソッドから`each`メソッドに戻す

yumiya-g committed:
Update: Promiseのエラーなしの処理を修正しました

yumiya-g committed:
Fix: データベース作成を関数化しました

yumiya-g committed:
Update: Promiseの処理を修正しました

- データベースの切断を`finally`メソッドで実行するよう修正
- エラーありの処理で、コールバック関数を見直しエラーハンドリングを追加

- Promiseでラップしている関数内の処理を外部関数に切り出し

yumiya-g committed:
Update: `Async/await`の処理を修正しました

yumiya-g committed:
Update: `sqlite3`を`functions.js`から読み込むように調整しました

yumiya-g committed:
Update: `Promise`の処理を修正しました

- 不要な引数`obj`を削除
- `this`が使えるよう、通常の関数定義で`console.log`を出力
- エラー処理内の書籍名を変更

yumiya-g committed:
Update: Async/awaitの処理を調整しました

- `promiseRun`を、for文で実行する処理に変更
- `promiseRun`と`promiseEach`に渡すクエリ、書籍名、コールバック関数を外部に切り出し
- DB切断する処理を`finally`メソッドで実行

cafedomancer commented on 03.asynchronous/functions.js:
ファイル名が抽象的すぎて何が定義されているファイルなのか分かりづらいです。

yumiya-g replied:
定義していた関数等を見直してファイル名を変更しました。

cafedomancer commented on 03.asynchronous/functions.js:
```diff
@@ -0,0 +1,67 @@
+import sqlite3 from "sqlite3";
+const createDB = () => new sqlite3.Database("database");
```
import とそれ以降のプログラムの間には空行を入れてください。

yumiya-g replied:
`import`のあとに改行を入れました。

cafedomancer commented on 03.asynchronous/functions.js:
```diff
@@ -0,0 +1,67 @@
+import sqlite3 from "sqlite3";
+const createDB = () => new sqlite3.Database("database");
```
インメモリのデータベースが使われていないです。

cafedomancer commented on 03.asynchronous/functions.js:
```diff
@@ -0,0 +1,67 @@
+export let db = createDB();
+export const recreateDB = () => {
+  db = createDB();
+};
```
他のプログラムで import されるもの (`db`) が、このファイルの別の関数 (`recreateDB`) で変更されるのは、挙動としてさすがに見通しが悪すぎるのでやめてください 🙅‍♂️ import する側のプログラムで `db` の中身がどう変化するのかが全く見通せないです。`new sqlite3.Database(...)` を必要に応じて普通に呼び出せばいいだけではないですかね？

yumiya-g replied:
上のコメントと合わせて`promise.js`でデータベースの接続処理を定義するよう修正しました🙏

cafedomancer commented on 03.asynchronous/functions.js:
```diff
@@ -0,0 +1,67 @@
+  db = createDB();
+};
+
+export const promiseRun = (query, params = [], callback = null) =>
```
デフォルト引数を何のために設定しているのか分からないので教えてほしいです 🙏 本当に必要ですか？ あと、元の関数を Promise で扱うようにしてるのにコールバック関数を受け取るのはおかしいような...。

yumiya-g replied:
このファイル内で定義した関数の引数の取り方を修正し、コールバック関数を受け取らないよう調整しました🙏

cafedomancer commented on 03.asynchronous/functions.js:
```diff
@@ -0,0 +1,67 @@
+  };
+};
+
+export const promiseEach = (query, ...args) =>
```
`promiseRun` では普通の引数を使っているのに、こちらでは可変長引数が使われていて一貫性がないように見えます。

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,84 @@
+  .finally(() => {
+    db.run(deleteTableQuery, () => {
+      db.close();
+    });
```
この部分だけ Promise になってないですよ。

yumiya-g replied:
`Promise`でラップする関数を定義して実行するよう修正しました。

yumiya-g committed:
fix: データベース接続処理をコード内で定義し、実行する関数に引数として渡すよう修正しました

- 外部ファイルに定義していたデータベース接続処理をコード内に定義して、変数`db`に代入
- 「エラーあり」「エラーなし」の処理を実行する関数を定義し、引数に`db`を渡す

yumiya-g committed:
fix: `Promise`でラップする関数にコールバック関数を渡す処理を修正しました

- 外部ファイルに定義していたデータベース接続処理をコード内に定義して、変数`db`に代入
- `Promise`でラップする関数に、引数として渡していたコールバック関数を削除
- `then`メソッドと`catch`メソッドで定義していた関数をアロー関数に変更
- `finally`メソッドで実行していたデータベース接続処理を`Promise`でラップ
- ファイル名が抽象的だったので修正しました

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,83 @@
+  });
+};
+
+registerDatabaseWithErrors(db);
```
関数の定義とトップレベルで実行している処理とが入り混じっていると見づらいです 🙅‍♂️ 関数の定義→トップレベルの処理の順にそれぞれまとめて行ってください。

yumiya-g replied:
> 関数の定義とトップレベルで実行している処理とが入り混じっている

定義と処理順を見直しました🙏

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,83 @@
+
+let db = new sqlite3.Database("database");
+
+const registerDatabase = (db) => {
```
この関数名だと「データベースに登録する」になってしまいます。データベースは登録するものではないですし、関数の処理内容的にも登録だけを行っているわけではないように思います :eyes:

yumiya-g replied:
関数を実行した結果に着目して命名を見直してみました。

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,72 @@
+      "スラスラ読める JavaScriptふりがなプログラミング",
+    ),
+  )
+  .then((obj) => {
```
`obj` だと何が格納されているのか分からないです。具体的な名前を付けてください 🙏 （`obj` のような抽象的な命名で良いケースはほぼ無いです）

yumiya-g replied:
`each`メソッドに合わせて`row`を採用してみました🙏
https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,72 @@
+      console.log(`ID: ${row.id}, タイトル: ${row.title}`);
+    });
+  })
+  .finally(() => promiseClose(db, deleteTableQuery));
```
テーブルの削除を finally で実行するのだと、テーブルの作成に失敗したときにもテーブルの削除をしようとするのでエラーになっちゃうと思います。

yumiya-g replied:
`finally`の前にテーブルを削除する処理を実行するよう調整しました。

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,72 @@
+  .catch((err) => {
+    console.error(err.message);
+    return promiseRun(db, insertTableQuery, null);
+  })
```
以下の部分がコールバックのプログラムと対応していないような。成功時と失敗時で異なる処理が実行されています。

```js
// 51-54 行目
  .then((obj) => {
    console.log(`ID: ${obj.lastID}`);
    return promiseRun(db, insertTableQuery, "初めてのJavaScript2");
  })

// (中略)

// 61-64 行目
  .catch((err) => {
    console.error(err.message);
    return promiseRun(db, insertTableQuery, null);
  })
```

コールバックのプログラムは以下のようになっていたはずです。成功時も失敗時もパラメータとして `null` を渡してクエリを発行していたはず。

```js
    db.run(
      insertTableWrongQuery,
      "スラスラ読める JavaScriptふりがなプログラミング",
      function (err) {
        if (err) {
          console.error(err.message);
        } else {
          console.log(`ID: ${this.lastID}`);
        }
        db.run(insertTableQuery, null, function (err) {
```

55-60 行目の以下の処理がこの間に挟まってるのもなんか変だと思います :eyes:

```js
// 55-60 行目
  .then((obj) => {
    console.log(`ID: ${obj.lastID}`);
    return promiseEach(db, selectTableQuery, (_err, row) => {
      console.log(`ID: ${row.id}, タイトル: ${row.title}`);
    });
  })
```

yumiya-g replied:
コールバック関数の処理に合わせて`then`と`catch`が実行される順番を調整しました🙏

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,72 @@
+      console.error(err.message);
+    });
+  })
+  .catch((err) => console.error(err.message))
```
アロー関数に戻り値がないときはそれを分かりやすくするために簡潔文体ではなくブロック文体で書いてください。https://github.com/airbnb/javascript?tab=readme-ov-file#arrows--implicit-return

yumiya-g replied:
`console.error`しか返していないのでブロック文体に修正しました。

cafedomancer commented on 03.asynchronous/promiseWrappedFunctions.js:
```diff
@@ -0,0 +1,28 @@
+
+export const promiseEach = (db, query, complete) =>
+  new Promise((resolve, reject) => {
+    db.each(query, complete, function (err) {
```
関数内で this を使う必要がない場合は function 式ではなくアロー関数を使ってください。https://github.com/airbnb/javascript?tab=readme-ov-file#arrows--use-them

yumiya-g replied:
関数内では条件分岐のみで`this`を使用していないので、アロー関数に修正しました。

cafedomancer commented on 03.asynchronous/promiseWrappedFunctions.js:
```diff
@@ -0,0 +1,28 @@
+      } else {
+        resolve();
+      }
+    });
```
これだと最初の 1 レコードを処理した段階ですぐに resolve されます。レコードの数を増やしてみるとちゃんと動いていないことが分かるはずです。

yumiya-g replied:
引数`params`を追加して、レコードの処理が終了した後に`complete`関数が実行されるよう調整しました🙏

cafedomancer commented on 03.asynchronous/promiseWrappedFunctions.js:
```diff
@@ -0,0 +1,28 @@
+  new Promise((resolve) => {
+    db.run(query, () => {
+      db.close(() => resolve());
+    });
```
1 つの関数で 2 つのことをやっています。元の close 関数には run の処理は含まれていないはずです。あとエラーが起きた場合が考慮されていない気がします。

yumiya-g replied:
`promiseClose`関数の処理を1つにして、エラーハンドリングを追加しました。

yumiya-g committed:
fix: トップレベルでの関数定義と処理実行順を調整しました

yumiya-g committed:
update: トップレベルに定義した関数の命名を見直しました

- 関数が実行されることと、エラーの有無がわかるように修正

yumiya-g committed:
fix: `each`メソッドの第2引数(param)に空配列を定義しました

- `each`メソッドの3番目の引数(callback)にコールバック関数を渡すため
- https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback

yumiya-g committed:
change: コールバック関数に合わせて`Promise`の関数定義を見直しました

- トップレベルで「エラーあり」「エラーなし」の関数名を定義
- 関数の実行処理順を調整

yumiya-g committed:
fix: `then`メソッドの引数名をrowに変更しました

- `each`メソッドのシグネチャーを参考
- https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

yumiya-g committed:
fix: データベースの切断処理にエラーハンドリングを追加しました

- `finally`メソッド内でテーブルの削除処理を除外して、1つの処理だけを行う

yumiya-g committed:
fix: テーブルの削除処理を`finally`メソッドの前で実行するよう調整しました

- `finally`でデータベース切断する直前に`then`メソッドでテーブルの削除を実行する

yumiya-g committed:
fix: `each`メソッドでコールバックが実行される処理を見直しました

- `promiseEach`の引数に空配列を渡す
- `each`メソッドの引数の順番を見直し、全レコードの処理完了後に`reject`または`resolve`を実行

yumiya-g committed:
fix: 'Promise'の処理を、コールバック(`callback.js`)の処理と合わせました

- `then`と`catch`の実行順番を修正
- 成功時と失敗時で同じ処理を実行

yumiya-g committed:
fix: 関数内で`this`を使用していないので、アロー関数に変更しました

yumiya-g committed:
fix: 引数に渡す文字列のタイポを修正しました

- `callback.js`に合わせて修正

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,83 @@
+          console.log(`ID: ${this.lastID}`);
+          db.each(
+            selectTableQuery,
+            [],
```
確認なんですが、これなんで空配列を渡すことになったんですかね？

yumiya-g replied:
ドキュメントに下記記載がありましたので`params`がない場合、明示的に空配列を渡したほうがよいかと考え、追加しました。

`In case you want to keep the callback as the 3rd parameter, you should set param to "[]" ( Empty Array ) as per [issue #116](https://github.com/mapbox/node-sqlite3/issues/116)`

https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback


ただ、`each`メソッドの説明文から、引数に関数を2つ定義した場合は`callback`と`complete`関数として扱われるようなので、空配列は削除します。

https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,83 @@
+  });
+};
+
+let db = new sqlite3.Database("database");
```
直ってないですよ。https://github.com/yumiya-g/js-practices/pull/4#discussion_r2178724948

というのと、`db` の定義だけ関数の外で行っている理由が分からなかったんですが、なぜですか？

yumiya-g replied:
どちらも修正漏れでした。インメモリのデータベースに修正し、`db`の定義を関数内で行うようにしました🙏

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,82 @@
+
+const runWithError = (db) => {
+  promiseRun(db, createTableQuery)
+    .then(() =>
+      promiseRun(
+        db,
+        insertTableWrongQuery,
+        "スラスラ読める JavaScriptふりがなプログラミング",
+      ),
+    )
+    .catch((err) => {
+      console.error(err.message);
+      return promiseRun(db, insertTableQuery, null);
+    })
```
コールバック版のプログラムでは正常時の処理があったのに、ここではないです。

yumiya-g replied:
`Promise`をラップする関数で処理が行われた後に、正常系を`then`で。異常系を`catch`で処理するようプロミスチェーンを修正しました。

yumiya-g committed:
fix: コールバック関数と`sqlite3.Database`オブジェクトの定義を修正しました

- 関数内部で変数`db`(`sqlite3.Database`オブジェクト)を定義して、関数外の変数に依存しないよう修正
- 関数定義から仮引数を削除

yumiya-g committed:
fix: インメモリのデータベースを使用するよう修正しました

yumiya-g committed:
fix: `each`メソッドの引数を調整しました

- `each`メソッドの第2引数に渡していた空配列を削除
- 引数に関数を2つ定義すると、1つ目がcallback、2つ目がcompleteとして扱われるため
  - https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

yumiya-g committed:
fix: コールバック関数の修正に合わせて、`Promise`を実行する関数の定義を修正しました

- 「エラーあり」「エラーなし」を実行する関数内部で変数db(sqlite3.Databaseオブジェクト)を定義
- 定義した関数から仮引数`db`を削除
- インメモリのデータベースを使用するよう修正

yumiya-g committed:
fix: `Promise`実行後の、正常系と異常系の処理を見直しました

- `Promise`をラップしている関数を実行後、`then`または`catch`で結果を出力表示
- 正常系または異常系の出力後、次の`Promise`をラップしている関数を実行している

yumiya-g committed:
fix: `promiseEach`から引数`params`を削除しました

- `each`メソッドに`params`を省略して関数を渡しても`callback`関数、`complete`関数と扱われるため
- https://github.com/TryGhost/node-sqlite3/wiki/API#eachsql--param---callback--complete

cafedomancer commented on 03.asynchronous/promiseWrappedFunctions.js:
```diff
@@ -0,0 +1,32 @@
+    });
+  });
+
+export const promiseEach = (db, query, callback) =>
```
元の each 関数はパラメータも受け取るようになっていたはずじゃないですかね？

yumiya-g replied:
引数`params`を受け取るようにして、パラメータが省略された場合の処理を追加しました🙏

cafedomancer commented on 03.asynchronous/promise.js:
```diff
@@ -0,0 +1,80 @@
+        "スラスラ読める JavaScriptふりがなプログラミング",
+      ),
+    )
+    .then((row) => {
```
ここで結果として得られるのは row ではないです。

yumiya-g replied:
`run`メソッドの実行結果が返されるので`result`に変更しました。

yumiya-g committed:
fix: `then`メソッドの引数名を修正しました

- `run`メソッドの実行結果が返されるので"result"に変更
- https://github.com/TryGhost/node-sqlite3/wiki/API#runsql--param---callback

yumiya-g committed:
fix: `each`メソッドに引数`params`を渡せるよう調整を行いました

- 仮引数に`params`を定義
- `params`が省略された際の条件分岐処理を追加

yumiya-g committed:
fix: Async/awaitのコードを`promise.js`に合わせて定義しなおしました

- importの定義を変更
- 関数名を`promise.js`と合わせて変更し、処理を削除
- `sqlite3.Database`をインメモリデータベースとして関数内で定義

yumiya-g committed:
fix: `Async`関数で「エラーなし」の処理を修正しました

- `await`式を使って逐次的に`Promise`を処理

yumiya-g committed:
fix: Async関数で「エラーあり」の処理を修正しました

- `await`式を使って逐次的に`Promise`を処理
- `try...catch`にてエラーハンドリング処理を追加

cafedomancer commented on 03.asynchronous/args.js:
このファイルはどこからも使われていないような。

yumiya-g replied:
不要になったファイルだったので削除をしました🙏

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,76 @@
+
+async function runWithoutError() {
+  const db = new sqlite3.Database(":memory:");
+  let result = await promiseRun(db, createTableQuery);
```
CREATE TABLE の結果は Promise のプログラムでは取得されてないですよ。この値自体もこのプログラムでは使われていないです。

yumiya-g replied:
`CREATE TABLE `によるテーブル作成の結果を、変数に代入をしないように修正をしました。

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,76 @@
+    console.log(`ID: ${row.id}, タイトル: ${row.title}`);
+  });
+  await promiseRun(db, deleteTableQuery);
+  await promiseClose(db);
```
Promise のプログラムでは close の処理は finally で実行されていたはずです。

yumiya-g replied:
`try...finally`文を使う処理に変更し、データベースの切断（`close`）処理を`finally`ブロックで行うよう修正しました。

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,76 @@
+    console.log(`ID: ${result.lastID}`);
+  } catch (err) {
+    console.error(err.message);
+  }
```
これだと例外がすべて握り潰されます。必要な例外だけ捕捉してください。Docs の注意点にも書いてあります。

yumiya-g replied:
`try...catch`文を見直して`catch`ブロックでエラーを`throw`する処理に変更。例外を握りつぶさないようにしました。

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,76 @@
+
+runWithoutError();
+
+await timers.setTimeout(100);
```
async/await のプログラムではこれは不要なはずです。Docs 注意点にもこれを使うようには書いてません。なぜかは考えましょう。

yumiya-g replied:
`await`式を使って`runWithoutError`関数と`runWithError `関数を逐次的に実行するよう調整し、`setTimeout`関数は削除しました。

yumiya-g committed:
delete: 利用されていない不要なファイルを削除しました

yumiya-g committed:
fix: テーブルを作成する`promiseRun`の結果を変数`result`に代入しない処理に修正しました

- `CREATE TABLE`でテーブルの作成だけを行っているため
  - 代入した変数`result`も使用されていない
- 「エラーなし」の処理では`try...catch`を使用しているので、変数を`const`で宣言をしている

yumiya-g committed:
fix: 「エラーなし」の`Async`関数のデータベース切断処理を見直しました

- `finally`ブロックでデータベース切断を実行

yumiya-g committed:
fix: Async関数「エラーあり」の例外処理を調整しました

- `promiseRun`関数を格納する変数`result`を`let`で定義しなおし
- `catch`メソッドでエラーをthrowして補足
- `finally`メソッドでテーブル削除とデータベース切断処理を実行する
  - エラー発生時を考慮して`finally`でデータベース削除を行うよう変更

yumiya-g committed:
fix: Async関数の「エラーあり」「エラーなし」の実行順を`await`で調整しました

- `timers/promises`モジュールと`setTimeOut`メソッドを削除
- トップレベルの`await`で`runWithoutError`と`runWithError`関数を逐次的に実行している
  - [非同期処理:Promise/Async Function · JavaScript Primer \#jsprimer](https://jsprimer.net/basic/async/#top-level-await-in-module)

yumiya-g committed:
fix: 例外処理でエラーが握りつぶさないよう`try...catch...finally`を調整しました

- 外側の`try`ブロック内に3回の`try...catch`を入れ子で配置
  - 内側の`catch`ブロックで`SQLITE`のエラーコードで条件分岐
  - 条件に一致しない例外があれば再スローする
- 内側の`catch`でスローされた例外を外側の`catch`で捕捉して、エラーログを出力

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+
+async function runWithoutError() {
+  const db = new sqlite3.Database(":memory:");
+  await promiseRun(db, createTableQuery);
```
(今回は起きませんが) ここでエラーが起きたときに finally の close が実行されないです。Promise 版のプログラムと異なる挙動のプログラムになっています。

yumiya-g replied:
エラーが発生した際に`finally`が実行されるよう`try`ブロック内に配置しなおしました🙏

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+      );
+      console.log(`ID: ${result.lastID}`);
+    } catch (err) {
+      if (err.code === "SQLITE_ERROR") {
```
これだと Error オブジェクト以外の以下のようなオブジェクトでも条件を満たしてしまいます。

```js
{ code: "SQLITE_ERROR" }
```

yumiya-g replied:
条件判定を`Error`オブジェクトかつ（特定の）エラーコードを持つ場合に修正しました。

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,92 @@
+  } catch (err) {
+    console.error(err.message);
+  } finally {
+    await promiseRun(db, deleteTableQuery);
```
以前にコメントしている内容が反映されていないです。https://github.com/yumiya-g/js-practices/pull/4#discussion_r2202064534

yumiya-g replied:
テーブルの削除処理を`finally`ブロックではなく、外側の`try`と`catch`ブロック内に配置しました。

yumiya-g committed:
fix: テーブル作成時のエラー発生を考慮しました

- テーブル作成の処理を`try`ブロック内に配置
- エラーが発生した際でも`finally`ブロックでデータベース切断処理を実行している

yumiya-g committed:
fix: テーブルを削除するタイミングを修正しました

- テーブル作成の処理を`try`ブロック内に配置
- 外側の`try`ブロック内でデータベースを削除する処理に修正
  - 同じく外側の`catch`ブロック内でデータベースを削除

yumiya-g committed:
fix: 「エラーあり」のエラー判定処理を修正しました

- 内部の`catch`ブロックで受け取った引数が`Error`オブジェクトか判定する処理を追加
- `Error`オブジェクトとエラーコードの出力内容で判定

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,95 @@
+    await promiseRun(db, deleteTableQuery);
+  } catch (err) {
+    console.error(err.message);
+    await promiseRun(db, deleteTableQuery);
```
DELETE TABLE の実行が 84 行目と重複しています。

yumiya-g replied:
`catch`ブロックを削除して、テーブルの削除処理が重複しないよう調整しました。

cafedomancer commented on 03.asynchronous/async.js:
```diff
@@ -0,0 +1,95 @@
+  } catch (err) {
+    console.error(err.message);
+    await promiseRun(db, deleteTableQuery);
+  } finally {
```
ここも例外の握り潰しが発生しています。例外を捕捉する必要がそもそも無さそうです。

yumiya-g replied:
想定外のエラー発生がした際に、例外が上部に伝搬するよう修正しました。

yumiya-g committed:
fix: 想定外の例外が発生した際のエラーハンドリングを調整しました

- `try...catch...finally`ブロックから`catch`ブロックを削除
  - テーブル作成時に想定外のエラーが発生した場合、テーブルの削除処理を実行せず上位に例外を伝搬させている
  - テーブル削除時に想定外のエラーが発生した場合、上位に例外を伝搬させている
